[clang-format] adds a space after not inside macros

mydeveloperday opened the pull request:
No need to add an extract space if merging the characters doesn't cause a problem, hence ) and ! as well as ( do not require the extraneous space.

Fixes #78166

mydeveloperday committed:
[clang-format] adds a space after not inside macros

No need to add an extract space if merging the characters
doesn't cause a problem, hence ) and ! as well as ( do not
require the extraneous space.

HazardyKnusperkeks commented:
I think this is kind of too specific. Whoever uses `not` will most likely not combine it with `!`. Also I agree with https://github.com/llvm/llvm-project/issues/78166#issuecomment-1892311219 that `WhitespaceSensitiveMacros` should be used.

owenca commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -4842,7 +4842,7 @@ bool TokenAnnotator::spaceRequiredBefore(const AnnotatedLine &Line,
   }
   if (Left.is(TT_UnaryOperator)) {
-    if (Right.isNot(tok::l_paren)) {
+    if (!Right.isOneOf(tok::r_paren, tok::l_paren, tok::exclaim)) {
```
> The only exception I see are identifiers and numeric literals (as they will combine the tokens together):
> 
> ```c++
> v(not x)
> v(not 1)
> v(not 123.f)
> ```

+1.

> Should this condition be something like `if (Right.isOneOf(tok::ident, /*... numeric literals*/))`? Or will this cause us to not add spaces in undesired places?
> 
> If we can avoid those spaces for only a limited set of tokens, I would probably limit this to `tok::lparen` and `tok::rparen`, the rest seems much less likely.

We should come up with a more general fix.

zygoloid replied:
For `not!`, I think adding a space is the right choice for a C++ code formatter: `not !b` should probably be formatted with a space, not as `not!b`, even though it's a pretty strange thing to write.

I wonder if this should instead be:
```suggestion
    if (!Right.isOneOf(tok::r_paren, tok::l_paren, tok::comma)) {
```
... with the `)` and `,` cases firing only in places where we can locally tell this isn't actually a `not` operator at all.

zygoloid replied:
Another data point: the [Python style guide](https://peps.python.org/pep-0008/#other-recommendations) also recommends always putting a space after `not`.

ilya-biryukov replied:
Adding a space makes total sense in code that **has an argument** to `not`.
But this case seems special because `V(not!)` or `V(not)` clearly can't be a full expression by themselves. So it feels preferable to not add any extra spaces.

owenca commented:
> I think this is kind of too specific.

+1.

> Also I agree with [#78166 (comment)](https://github.com/llvm/llvm-project/issues/78166#issuecomment-1892311219) that `WhitespaceSensitiveMacros` should be used.

I don't think we should require that the option be used.

mydeveloperday commented:
> > I think this is kind of too specific.
> 
> +1.
> 
> > Also I agree with [#78166 (comment)](https://github.com/llvm/llvm-project/issues/78166#issuecomment-1892311219) that `WhitespaceSensitiveMacros` should be used.
> 
> I don't think we should require that the option be used.

+1 to that too! (even though it works)

owenca commented:
This should work:
```
--- a/clang/lib/Format/TokenAnnotator.cpp
+++ b/clang/lib/Format/TokenAnnotator.cpp
@@ -4402,9 +4402,17 @@ bool TokenAnnotator::spaceRequiredBefore(const AnnotatedLine &Line,
   if (Left.Finalized)
     return Right.hasWhitespaceBefore();
 
-  // Never ever merge two words.
-  if (Keywords.isWordLike(Right) && Keywords.isWordLike(Left))
-    return true;
+  if (Style.isVerilog()) {
+    if (Keywords.isWordLike(Left) && Keywords.isWordLike(Right))
+      return true;
+  } else {
+    auto IsWordOrNumber = [](const auto &Tok) {
+      return Tok.Tok.getIdentifierInfo() || Tok.is(tok::numeric_constant);
+    };
+    // Never ever merge two words/numbers.
+    if (IsWordOrNumber(Left) && IsWordOrNumber(Right))
+      return true;
+  }
 
   // Leave a space between * and /* to avoid C4138 `comment end` found outside
   // of comment.
@@ -4842,21 +4850,11 @@ bool TokenAnnotator::spaceRequiredBefore(const AnnotatedLine &Line,
     return true;
   }
   if (Left.is(TT_UnaryOperator)) {
-    if (Right.isNot(tok::l_paren)) {
-      // The alternative operators for ~ and ! are "compl" and "not".
-      // If they are used instead, we do not want to combine them with
-      // the token to the right, unless that is a left paren.
-      if (Left.is(tok::exclaim) && Left.TokenText == "not")
-        return true;
-      if (Left.is(tok::tilde) && Left.TokenText == "compl")
-        return true;
-      // Lambda captures allow for a lone &, so "&]" needs to be properly
-      // handled.
-      if (Left.is(tok::amp) && Right.is(tok::r_square))
-        return Style.SpacesInSquareBrackets;
-    }
-    return (Style.SpaceAfterLogicalNot && Left.is(tok::exclaim)) ||
-           Right.is(TT_BinaryOperator);
+    // Lambda captures allow for a lone &, so "&]" needs to be properly
+    // handled.
+    if (Left.is(tok::amp) && Right.is(tok::r_square))
+      return Style.SpacesInSquareBrackets;
+    return Style.SpaceAfterLogicalNot && Left.is(tok::exclaim);
   }
 
   // If the next token is a binary operator or a selector name, we have
```

mydeveloperday committed:
[clang-format] adds a space after not inside macros

No need to add an extract space if merging the characters
doesn't cause a problem, hence ) and ! as well as ( do not
require the extraneous space.

mydeveloperday commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -24160,6 +24160,14 @@ TEST_F(FormatTest, AlternativeOperators) {
+  verifyFormat("Symbol(not, None)");
+  verifyFormat("Symbol(not!, None)");
+
+  verifyFormat("assert(!\"fail\")");
+  verifyFormat("assert(not\"fail\")");
```
I've seen this used extensively, I'm not sure it triggers a compile warning.

assert(!"fail")

owenca replied:
You're right!

ilya-biryukov commented:
I just wanted to bring up that using `WhitespaceSensitiveMacros` here is a little tricky because the macro name one would need to put there is `V` (the name is taken directly from the code that hit this).
Having a more sophisticated option to say "parameter `V` of macro `MyComplicatedMacro` is a string-sensitive macro" should work better, but seems like an overkill.
